src/environment/SCRF.cpp

```diff
@@ -242,63 +227,41 @@ void SCRF::printConvergenceRow(int i, double norm, double update, double time) c
 
 mrcpp::ComplexFunction &SCRF::setup(double prec, const OrbitalVector_p &Phi) {
     this->apply_prec = prec;
-    computeDensities(*Phi);
+    Density rho_tot(false);
+    computeDensities(*Phi, rho_tot);
     Timer t_vac;
     mrcpp::ComplexFunction V_vac;
     V_vac.alloc(NUMBER::Real);
-    mrcpp::apply(this->apply_prec, V_vac.real(), *poisson, this->rho_tot.real());
+    mrcpp::apply(this->apply_prec, V_vac.real(), *poisson, rho_tot.real());
+    rho_tot.free(NUMBER::Real);
     print_utils::qmfunction(3, "Vacuum potential", V_vac, t_vac);
 
     // set up the zero-th iteration potential and gamma, so the first iteration gamma and potentials can be made
 
     Timer t_gamma;
-    if ((not this->Vr_n.hasReal()) or (not this->gamma_n.hasReal())) {
+    if (not this->Vr_n.hasReal()) {
         mrcpp::ComplexFunction gamma_0;
         mrcpp::ComplexFunction V_tot;
         computeGamma(V_vac, gamma_0);
-        this->Vr_n = solvePoissonEquation(gamma_0);
-        mrcpp::cplxfunc::add(V_tot, 1.0, V_vac, 1.0, this->Vr_n, -1.0);
-        computeGamma(V_tot, this->gamma_n);
+        this->Vr_n = solvePoissonEquation(gamma_0, *Phi);
     }
 
     // update the potential/gamma before doing anything with them
 
-    if (accelerate_Vr) {
-        mrcpp::ComplexFunction temp_Vr_n;
-        mrcpp::cplxfunc::add(temp_Vr_n, 1.0, this->Vr_n, 1.0, this->dVr_n, -1.0);
-        mrcpp::cplxfunc::deep_copy(this->Vr_n, temp_Vr_n);
-        temp_Vr_n.free(NUMBER::Real);
-        mrcpp::ComplexFunction V_tot;
-        mrcpp::cplxfunc::add(V_tot, 1.0, this->Vr_n, 1.0, V_vac, -1.0);
-        resetComplexFunction(this->gamma_n);
-        computeGamma(V_tot, this->gamma_n);
-    } else {
-        mrcpp::ComplexFunction temp_gamma_n;
-        mrcpp::cplxfunc::add(temp_gamma_n, 1.0, this->gamma_n, 1.0, this->dgamma_n, -1.0);
-        mrcpp::cplxfunc::deep_copy(this->gamma_n, temp_gamma_n);
-        temp_gamma_n.free(NUMBER::Real);
-    }
-    print_utils::qmfunction(3, "Initial gamma", this->gamma_n, t_gamma);
-
     Timer t_scrf;
-    nestedSCRF(V_vac);
+    nestedSCRF(V_vac, Phi);
     print_utils::qmfunction(3, "Reaction potential", this->Vr_n, t_scrf);
     return this->Vr_n;
 }
 
-double SCRF::getNuclearEnergy() {
-    return mrcpp::cplxfunc::dot(this->rho_nuc, this->Vr_n).real();
-}
-
-double SCRF::getElectronicEnergy() {
-    mrcpp::ComplexFunction rho_el;
-    rho_el.alloc(NUMBER::Real);
-    mrcpp::cplxfunc::add(rho_el, 1.0, this->rho_tot, -1.0, this->rho_nuc, -1.0);
-    return mrcpp::cplxfunc::dot(rho_el, this->Vr_n).real();
-}
-
-double SCRF::getTotalEnergy() {
-    return mrcpp::cplxfunc::dot(this->rho_tot, this->Vr_n).real();
+void SCRF::computeEnergies(mrchem::OrbitalVector &Phi) {
```

Comment:
I think the orbitals can be passed by `const`-reference.
```suggestion
void SCRF::computeEnergies(const mrchem::OrbitalVector &Phi) {
```

Reply:
the `density::compute` method discards `const` qualifier, according to the compile error. I think the `cplxfunc::add` and `multiply` also do this  for the input `ComplexFunction`.

Reply:
maybe pass the electronic density directly?

Reply:
> maybe pass the electronic density directly?

that's indeed what I ended up doing in #456 😅

Reply:
Ok, I am using a const reference to the density now